src/features/genetics-gd/nodes/entity_gd.gd

```diff
@@ -14,24 +16,86 @@ var paused : bool :
 		
 
 var traits := {}
+
+## each element should be a list (length 2 = diploid) of traits
+## loci should be the key
+var genotype := {} 
+
+## each element should be a loci key with some number of traits corresponding to the maximum dominance value
+var phenotype := {}
+
+## how many genes in each loci (humans and our creatures are diploid)
+## eventually, this should be configurable
+var ploidy : int = 2
+
+signal traits_changed
  
 func _ready() -> void:
-	for t in self.initial_traits:
-		self.add_trait(t)
+	for l:Loci in self.initial_genotype:
+		for t_index in range(len(l.traits)): 
+			self.add_trait(l.traits[t_index], t_index)
+
+var none_scene := preload("res://features/genetics-gd/nodes/entity_traits/trait_none.tscn")
+func create_loci(loci: String) -> void:
+	genotype[loci] = []
+	for i in range(ploidy):
+		var none : TraitNone = none_scene.instantiate()
+		none.loci = loci
+		genotype[loci].append(none)
 
+func add_trait(new_trait: PackedScene, loci_index: int = 0) -> void:
+	if new_trait.can_instantiate():
+		var _new_trait := new_trait.instantiate()
+		_add_trait(_new_trait, loci_index)
 
-func add_trait(new_trait: PackedScene) -> void:
-	var _new_trait := new_trait.instantiate()
+func _add_trait(_new_trait: TraitBase, loci_index: int = 0) -> void:
 	assert(_new_trait is TraitBase)
-	self.traits[_new_trait.unique_trait_name] = _new_trait
-	add_child(_new_trait)
-	
-func remove_trait(unique_trait_name: String) -> bool:
-	var _trait: TraitBase = self.traits.get(unique_trait_name)
-	if _trait != null:
-		return false
+	if _new_trait.loci in genotype and _new_trait.unique_trait_name != genotype[_new_trait.loci][loci_index].unique_trait_name:
+		genotype[_new_trait.loci][loci_index] = _new_trait
+	elif _new_trait.loci not in genotype:
+		create_loci(_new_trait.loci)
+		genotype[_new_trait.loci][loci_index] = _new_trait
+		phenotype[_new_trait.loci] = [] as Array[TraitBase]
 		
-	self.traits.erase(unique_trait_name)
-	remove_child(_trait)
+	var dominant := dominant_trait_at_loci(_new_trait.loci)
+	for i:TraitBase in phenotype[_new_trait.loci]:
+		if trait_is_in(_new_trait, dominant):
+			phenotype[_new_trait.loci].erase(i)
+			traits.erase(i.unique_trait_name)
+			i.queue_free()
+	for i:TraitBase in dominant:
+		if trait_is_in(_new_trait, phenotype[_new_trait.loci]) and not i is TraitNone:
+
+			phenotype[_new_trait.loci].append(i)
+			traits[_new_trait.unique_trait_name] = _new_trait
+			self.add_child(i)
+	traits_changed.emit()
+
+func trait_is_in(t: TraitBase, arr: Array[TraitBase]) -> bool:
+	var _trait_is_in : bool = false
+		# only unique traits get added to phenotype
+	for n : TraitBase in arr:
+		if n.unique_trait_name == t.unique_trait_name:
+			_trait_is_in = true
+	return _trait_is_in
```

Comment:
I believe this could be simplified/optimized so that it does not run through the whole loop every time (unless necessary).

```suggestion
		# only unique traits get added to phenotype
	for n : TraitBase in arr:
		if n.unique_trait_name == t.unique_trait_name:
			return true
	return false
```